_posts/2021-04-02-being-productive-with-typescript.md

```diff
@@ -0,0 +1,477 @@
+- ~~Code hinting~~
+- ~~Compile-time errors~~
+
+```typescript
+type DateTimeString = string; //yyyy-MM-ddThh:mm:ss.sssZ
+type DateString = string; //yyyy-MM-dd
+type Seconds = number;
+
+interface CalenderItemDTO {
+ duration?: Seconds,
+ startTime: DateTimeString,
+ dueDate?: DateString,
+}
+
+interface CalenderItem {
+ duration?: Seconds,
+ startTime: Date,
+ dueDate?: Date,
+}
+
+function mapCalenderItemFromDTO(item: CalenderItemDTO): CalenderItemDTO {
+ return { ...item,
+   startTime: new Date(item.startTime),
+   dueDate: item.dueDate 
+     ? new Date(item.dueDate) 
+     : undefined
+ }
+}
+
+function mapCalenderItemToDTO(item: CalenderItem): CalenderItem {
+ return { ...item,
+   startTime: item.startTime.toISOString(),
+   dueDate: item.dueDate
+     ? format(item.dueDate, 'yyyy-MM-dd') 
+     : undefined
+ }
+}
+```
+The code became a lot bigger. 
+Some people grade code on the amount of lines written to achieve a goal.
+While it is fun to think about shorter code, I would always argue that maintainability is key in production code.
+
+Let us begin with the types. 
+Typescript allows us to create [Type Aliases](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases){:target="_blank" rel="noopener noreferrer"}. 
+Its only purpose, is making your code more readable. 
+It does not improve error descriptions, neither does it add typesafety, but it improves your variable declarations.
+The next developer will know immediately that `duration` is in seconds, for example. The same goes for `startTime` and `dueDate`.
+
+Next significant change is the DTO type. 
+A [Data Transfer Object](https://en.wikipedia.org/wiki/Data_transfer_object){:target="_blank" rel="noopener noreferrer"} can be used for communication between systems.
+It can be seen as a tiny anti-corruption layer, because the internal code does not need to know about communication with an external system.
+The direct benefit is that we always know which types we are using and do not need explicit typecasting.
+A keen eye might also notice the absence of [union types](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-types){:target="_blank" rel="noopener noreferrer"}.
+[Union types](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-types){:target="_blank" rel="noopener noreferrer"} tend to break the code hinting. 
+It is better to avoid them when possible.
+
+Let's evaluate the code using the Typescript features:
+- Readability
+- Code hinting
+- Compile-time errors
+
+It seems perfect, but we are violating the DRY principle now.
+Every time the contract changes, we will have to adapt 2 interfaces. 
+Even when there is no functional impact.
+Remember that Typescript interfaces are not transpiled, so I gave myself some extra work for _undefined_.
+Puns aside, there is an easy improvement we can do.
+
+```typescript
+interface CalenderItemDTO {
+ title: string,
+ location?: Location,
```

Comment:
You probably added `title` and `location` here to show that you just have to add them to 1 place, but actually that wasn't clear to me (it still is just my best guess of why they are all of the sudden here)

Reply:
added a clearer explanation in the text below